pkg/profiles/rule_validator.go

```diff
@@ -67,36 +67,55 @@ func (r *RuleValidator) ValidateParamsAgainstSchema(params map[string]any) error
 	if r.paramSchema == nil {
 		return nil
 	}
-
 	if err := validateAgainstSchema(r.paramSchema, params); err != nil {
 		return &RuleValidationError{
 			RuleType: r.ruleTypeName,
 			Err:      err.Error(),
 		}
 	}
-
+	applyDefaults(r.paramSchema, params)
 	return nil
 }
 
-func validateAgainstSchema(schema *gojsonschema.Schema, obj map[string]any) error {
-	documentLoader := gojsonschema.NewGoLoader(obj)
-	result, err := schema.Validate(documentLoader)
-	if err != nil {
-		return fmt.Errorf("cannot validate json schema: %s", err)
+func compileSchema(schemaData interface{}) (*jsonschema.Schema, error) {
+	compiler := jsonschema.NewCompiler()
+	if err := compiler.AddResource("schema.json", schemaData); err != nil {
+		return nil, fmt.Errorf("invalid schema: %w", err)
 	}
+	return compiler.Compile("schema.json")
+}
 
-	if !result.Valid() {
-		return buildValidationError(result.Errors())
+func validateAgainstSchema(schema *jsonschema.Schema, obj map[string]any) error {
+	if err := schema.Validate(obj); err != nil {
+		return buildValidationError(err.(*jsonschema.ValidationError).Causes)
 	}
-
 	return nil
 }
 
-func buildValidationError(errs []gojsonschema.ResultError) error {
+func buildValidationError(errs []*jsonschema.ValidationError) error {
 	problems := make([]string, 0, len(errs))
 	for _, desc := range errs {
-		problems = append(problems, desc.String())
+		problems = append(problems, desc.Error())
 	}
-
 	return fmt.Errorf("invalid json schema: %s", strings.TrimSpace(strings.Join(problems, "\n")))
 }
+
+// applyDefaults recursively applies default values from the schema to the object.
+func applyDefaults(schema *jsonschema.Schema, obj map[string]any) {
```

Comment:
I wish the library had this function, but it looks like it doesn't.  😞

Reply:
No library I found has one :cry: